bool is_empty
bool fence_open